File: tests/src/org.playframework.anorm/anorm_3/2.8.1/src/test/scala/org_playframework_anorm/anorm_3/SelectableTest.scala
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_playframework_anorm.anorm_3

import java.time.Month

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

import scala.reflect.Selectable

class SelectableTest {
  @Test
  def selectsPublicFieldDynamically(): Unit = {
    val selectable: Selectable = Selectable.reflectiveSelectable(Month.JANUARY)

    val selected: Any = selectable.selectDynamic("DECEMBER")

    assertThat(selected.asInstanceOf[Object]).isEqualTo(Month.DECEMBER)
  }

  @Test
  def invokesPublicMethodDynamically(): Unit = {
    val selectable: Selectable = Selectable.reflectiveSelectable("playframework")
    val selected: Any = selectable.applyDynamic("substring", classOf[Int], classOf[Int])(
      Integer.valueOf(0),
      Integer.valueOf(4)
    )

    assertThat(selected.asInstanceOf[Object]).isEqualTo("play")
  }
}

File: tests/src/org.scala-lang/scala3-library_3/3.0.0/src/test/java/org_scala_lang/scala3_library_3/SelectableTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_scala_lang.scala3_library_3;

import java.util.List;

import org.junit.jupiter.api.Test;

import scala.collection.immutable.Seq;
import scala.collection.immutable.Seq$;
import scala.jdk.javaapi.CollectionConverters;
import scala.reflect.Selectable;
import scala.reflect.Selectable$;

import static org.assertj.core.api.Assertions.assertThat;

public class SelectableTest {
    @Test
    void selectDynamicReadsPublicField() {
        Selectable selectable = Selectable$.MODULE$.reflectiveSelectable(Integer.valueOf(0));

        Object value = selectable.selectDynamic("MAX_VALUE");

        assertThat(value).isEqualTo(Integer.MAX_VALUE);
    }

    @Test
    void applyDynamicInvokesPublicMethod() {
        Selectable selectable = Selectable$.MODULE$.reflectiveSelectable("native-image");
        Seq<Class<?>> parameterTypes = Seq$.MODULE$.from(
                CollectionConverters.asScala(List.<Class<?>>of(int.class, int.class)));
        Seq<Object> arguments = Seq$.MODULE$.from(
                CollectionConverters.asScala(List.<Object>of(0, 6)));

        Object value = selectable.applyDynamic("substring", parameterTypes, arguments);

        assertThat(value).isEqualTo("native");
    }
}
